Cannot parse extern type declarations

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
extern "C" {
    type Opaque;
}
```

I expected to see this happen: no error at the parsing stage - an error when feature gating.

`rustc test.rs -Z parse-only` does not throw errors

Instead, this happened:
```rust
test.rs:2:5: error: unrecognised token ‘type’ in extern block item declaration
    2 |     type Opaque;
      |     ^~~~
test.rs:2:5: error: failed to parse external item despite not reaching end of extern block
test.rs:3:1: error: failed to parse item in crate
    3 | }
      | ^

Analyzing compilation unit

Time variable                                   usr           sys          wall           GGC
 phase setup                        :   0.00 (  0%)   0.00 (  0%)   0.01 (100%)   137k ( 95%)
 TOTAL                              :   0.00          0.00          0.01          145k
Extra diagnostic checks enabled; compiler may run slowly.
Configure with --enable-checking=release to disable checks.
```

### Meta

- What version of Rust GCC were you using: 0d5fe7b9aee

Also ties in with #1579 (feature `extern_types`)

--- a/gcc/rust/parse/rust-parse-impl.h
+++ b/gcc/rust/parse/rust-parse-impl.h
@@ -3983,6 +3983,34 @@ Parser<ManagedTokenSource>::parse_lifetime ()
     }
 }
 
+template <typename ManagedTokenSource>
+std::unique_ptr<AST::ExternalTypeItem>
+Parser<ManagedTokenSource>::parse_external_type_item (AST::Visibility vis,
+						      AST::AttrVec outer_attrs)
+{
+  Location locus = lexer.peek_token ()->get_locus ();
+  skip_token (TYPE);
+
+  const_TokenPtr alias_name_tok = expect_token (IDENTIFIER);
+  if (alias_name_tok == nullptr)
+    {
+      Error error (lexer.peek_token ()->get_locus (),
+		   "could not parse identifier in external opaque type");
+      add_error (std::move (error));
+
+      skip_after_semicolon ();
+      return nullptr;
+    }
+
+  if (!skip_token (SEMICOLON))
+    return nullptr;
+
+  return std::unique_ptr<AST::ExternalTypeItem> (
+    new AST::ExternalTypeItem (std::move (alias_name_tok->get_str ()),
+			       std::move (vis), std::move (outer_attrs),
+			       std::move (locus)));
+}
+
 // Parses a "type alias" (typedef) item.
 template <typename ManagedTokenSource>
 std::unique_ptr<AST::TypeAlias>
@@ -6044,6 +6072,9 @@ Parser<ManagedTokenSource>::parse_external_item ()
 	    std::move (variadic_attrs), std::move (vis),
 	    std::move (outer_attrs), locus));
       }
+    case TYPE:
+      return parse_external_type_item (std::move (vis),
+				       std::move (outer_attrs));
     default:
       // error
       add_error (

--- a/gcc/rust/parse/rust-parse.h
+++ b/gcc/rust/parse/rust-parse.h
@@ -273,6 +273,8 @@ private:
   template <typename EndTokenPred>
   std::vector<AST::Lifetime> parse_lifetime_bounds (EndTokenPred is_end_token);
   AST::Lifetime parse_lifetime ();
+  std::unique_ptr<AST::ExternalTypeItem>
+  parse_external_type_item (AST::Visibility vis, AST::AttrVec outer_attrs);
   std::unique_ptr<AST::TypeAlias> parse_type_alias (AST::Visibility vis,
 						    AST::AttrVec outer_attrs);
   std::unique_ptr<AST::Struct> parse_struct (AST::Visibility vis,

--- a/gcc/testsuite/rust/compile/extern_type_item.rs
+++ b/gcc/testsuite/rust/compile/extern_type_item.rs
@@ -1,0 +1,5 @@
+// { dg-additional-options "-fsyntax-only" }
+
+extern "C" {
+    type F;
+}

--- a/gcc/testsuite/rust/compile/extern_type_item_missing_semi.rs
+++ b/gcc/testsuite/rust/compile/extern_type_item_missing_semi.rs
@@ -1,0 +1,7 @@
+// { dg-additional-options "-fsyntax-only" }
+
+extern "C" {
+    type F;
+    type E // { dg-error "failed to parse" }
+} // { dg-error "expecting" }
+// { dg-error "failed to parse item in crate" ""  { target *-*-* } .-1 }

